lib/smooth/core/io/spi/Master.cpp

```diff
@@ -53,71 +56,92 @@ namespace smooth::core::io::spi
         bus_config.quadhd_io_num = quadhd_io_num;
         bus_config.max_transfer_sz = transfer_size;
 
-        spi_bus_initialized = false;
+        bool initialized = false;
 
         if (spi_host == VSPI_HOST)
         {
-            spi_bus_initialized = do_vspi_initialization();
+            initialized = do_intitialization(VSPI_HOST, vspi_initialized, vspi_initialized_count, vspi_host_str);
         }
 
         if (spi_host == HSPI_HOST)
         {
-            spi_bus_initialized = do_hspi_initialization();
+            initialized = do_intitialization(HSPI_HOST, hspi_initialized, hspi_initialized_count, hspi_host_str);
         }
 
-        return spi_bus_initialized;
+        return initialized;
     }
 
-    bool Master::do_vspi_initialization()
+    bool Master::do_intitialization(spi_host_device_t host,
+                                    bool& initialized,
+                                    uint8_t& initialized_count,
+                                    const char* spi_host_str)
     {
-        if (!vspi_initialized)
+        if (!initialized)
         {
-            vspi_initialized = spi_bus_initialize(spi_host, &bus_config, dma_channel) == ESP_OK;
+            initialized = spi_bus_initialize(host, &bus_config, dma_channel) == ESP_OK;
 
-            if (!vspi_initialized)
+            if (!initialized)
             {
-                Log::error(log_tag, "VSPI_HOST Initialization failed");
+                Log::error(log_tag, "{} Initialization has failed", spi_host_str);
             }
             else
             {
-                Log::info(log_tag, "VSPI initialized, Host {}, DMA {}", spi_host, dma_channel);
+                Log::info(log_tag, "{} has been initialized using DMA channel {}", spi_host_str, dma_channel);
             }
         }
 
-        return vspi_initialized;
-    }
-
-    bool Master::do_hspi_initialization()
-    {
-        if (!hspi_initialized)
+        // if initialized then increment initialized_count so we know how many spi-devices are on this spi-bus
+        if (initialized)
         {
-            hspi_initialized = spi_bus_initialize(spi_host, &bus_config, dma_channel) == ESP_OK;
-
-            if (!hspi_initialized)
-            {
-                Log::error(log_tag, "HSPI_HOST Initialization failed");
-            }
-            else
-            {
-                Log::info(log_tag, "HSPI initialized, Host {}, DMA {}", spi_host, dma_channel);
-            }
+            initialized_count++;
```

Comment:
This member needs to be protected to support multi threading. Both init and de-init paths must ensure they lock the mutex so that only one path is run at a time.

Reply:
Something is strange here.  Can you look at the actual file Master.cpp and Master.h because it looks like changes tracked by git are not correct.

It was probably something I did with git.  I followed theses steps https://www.tomasbeuzen.com/post/git-fork-branch-pull/
and got up to step 7 and opened this pull request.  When you requested changes I repeated step 6    $git push origin spi-master-updates but I'm not sure if this is correct.

Reply:
Also CMakeLists.txt looks correct when I open the file Smooth/CMakeLists.txt in spi-master-updates.
So not sure what to do???

Reply:
I will wait for your comment from my comments above before I correct SPISDCard.cpp just in case there are other changes.

Reply:
It's OK, I must have missed it, the lock is there.

The selected project is however actually not correct, at least not what you have pushed to git:
https://github.com/enelson1001/Smooth/blob/spi-master-updates/CMakeLists.txt#L25

Reply:
OK this last PUSH should complete your requests.  Hopefully next pull request I do will be better.